executor/partition.go

```diff
@@ -0,0 +1,371 @@
+//          |                 |    |            |
+//          |          +------o----+            |
+//          |          |      +-----------------+-----+
+//          |          |                              |
+//          |      +---+------------+------------+----+-----------+
+//          |      |             Partition  Execution             |
+//          |      +--------------+-+------------+-+--------------+
+//          |                             ^
+//          |                             |
+//          |             +---------------v-----------------+
+//          +---------->  |    fetch data from DataSource   |
+//                        +---------------------------------+
+//
+////////////////////////////////////////////////////////////////////////////////////////
+type PartitionExec struct {
+	baseExecutor
+	concurrency int
+	workers     []*partitionWorker
+
+	prepared bool
+	executed bool
+
+	splitter   partitionSplitter
+	dataSource Executor
+
+	finishCh chan struct{}
+	outputCh chan *partitionOutput
+}
+
+type partitionOutput struct {
+	chk        *chunk.Chunk
+	err        error
+	giveBackCh chan *chunk.Chunk
+}
+
+// Open implements the Executor Open interface
+func (e *PartitionExec) Open(ctx context.Context) error {
+	if err := e.dataSource.Open(ctx); err != nil {
+		return err
+	}
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+
+	e.prepared = false
+	e.finishCh = make(chan struct{}, 1)
+	e.outputCh = make(chan *partitionOutput, e.concurrency)
+
+	for _, w := range e.workers {
+		w.finishCh = e.finishCh
+
+		w.inputCh = make(chan *chunk.Chunk, 1)
+		w.inputHolderCh = make(chan *chunk.Chunk, 1)
+		w.outputCh = e.outputCh
+		w.outputHolderCh = make(chan *chunk.Chunk, 1)
+
+		if err := w.childExec.Open(ctx); err != nil {
+			return err
+		}
+
+		w.inputHolderCh <- newFirstChunk(e.dataSource)
+		w.outputHolderCh <- newFirstChunk(e)
+	}
+
+	return nil
+}
+
+// Close implements the Executor Close interface
+func (e *PartitionExec) Close() error {
+	if !e.prepared {
+		for _, w := range e.workers {
+			close(w.inputHolderCh)
+			close(w.inputCh)
+			close(w.outputHolderCh)
+		}
+		close(e.outputCh)
+	}
+	close(e.finishCh)
+	for _, w := range e.workers {
+		for range w.inputCh {
+		}
+	}
+	for range e.outputCh {
+	}
+	e.executed = false
+
+	if e.runtimeStats != nil {
+		e.runtimeStats.SetConcurrencyInfo("PartitionConcurrency", e.concurrency)
+	}
+
+	err := e.dataSource.Close()
+	err1 := e.baseExecutor.Close()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(err1)
+}
+
+func (e *PartitionExec) prepare4ParallelExec(ctx context.Context) {
+	go e.execDataFetcherThread(ctx)
+
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(e.workers))
+	for _, w := range e.workers {
+		go w.run(ctx, waitGroup)
+	}
+
+	go e.waitWorkerAndCloseOutput(waitGroup)
+}
+
+func (e *PartitionExec) waitWorkerAndCloseOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
```

Comment:
To ensure all goroutines exited, should we also need wait this `waitGroup` in `Close` method?

Reply:
I think we don't need it.
`PartitionExec.Close` waits for the close of `outputCh`, which is only closed by `waitWorkerAndCloseOutput` after `waitGroup.Wait()`.